node/src/env0-deploy-utils.js

```diff
@@ -50,49 +50,95 @@ class DeployUtils {
     }
 
     console.log(`setting the following configuration: ${JSON.stringify(configuration)}`);
-    await apiClient.callApi('post', 'configuration', {data: {...configuration, projectId: undefined}});
+    await this.apiClient.callApi('post', 'configuration', {data: {...configuration, projectId: undefined}});
   }
 
   async deployEnvironment(environment, blueprintRevision, blueprintId) {
-    console.log(`starting to deploy environmentId: ${environment.id}, blueprintId: ${blueprintId}`);
-    const deployment = await apiClient.callApi('post', `environments/${environment.id}/deployments`,
-        {data: {blueprintId, blueprintRevision}});
+    await this.waitForEnvironment(environment.id);
 
-    console.log(`Started deployment ${deployment.id}`);
+    return await this.apiClient.callApi('post', `environments/${environment.id}/deployments`, {data: {blueprintId, blueprintRevision}});
   }
 
   async destroyEnvironment(environment) {
-    console.log(`Starting to destroy environmentId: ${environment.id}`);
-    const deployment = await apiClient.callApi('post', `environments/${environment.id}/destroy`);
-    console.log(`Started destroy ${deployment.id}`);
+    await this.waitForEnvironment(environment.id);
+
+    return await this.apiClient.callApi('post', `environments/${environment.id}/destroy`);
   }
 
+  async writeDeploymentStepLog(deploymentLogId, stepName) {
+    let hasMoreLogs;
 
-  async pollEnvironmentStatus(environmentId) {
-    const deployEndStatuses = ['CREATED', 'ACTIVE', 'INACTIVE', 'FAILED', 'TIMEOUT'];
-    const maxRetryNumber = 90; //waiting for 15 minutes (90 * 10 seconds)
-    let retryCount = 0;
+    do {
+      const { events, hasMoreLogs: currentHasMoreLogs } = await this.apiClient.callApi('get', `deployments/${deploymentLogId}/steps/${stepName}/log`);
```

Comment:
You need to add `startTime` if you want to do proper paging here.
Like this you will get duplicate logs.

You can see `frontend/src/stores/deployment-steps.store.ts` to understand how the frontend does this.
We don't need to 'merge' the logs from each call, but we only need the delta each time.

Reply:
8e591b7
d0f4500